surveyor-brick/src/Surveyor/Brick.hs

```diff
@@ -312,18 +312,90 @@ stateFromContext :: forall arch s p sym ext rtp f a st fs
+              funcViewerCallback rep fh b = do
+                C.emitEvent chan (C.PushContext (C.archNonce ares) fh rep b)
+                C.emitEvent chan (C.ViewBlock (C.archNonce ares) rep)
+          let funcViewers = MapF.Pair C.BaseRepr (FV.functionViewer (funcViewerCallback C.BaseRepr) FunctionCFGViewer C.BaseRepr)
+                            : [ MapF.Pair rep (FV.functionViewer (funcViewerCallback rep) FunctionCFGViewer rep)
+                              | C.SomeIRRepr rep <- C.alternativeIRs (Proxy @(arch, s))
+                              ]
+          let uiState = SBE.BrickUIState { SBE.sBlockSelector = BS.emptyBlockSelector
+                                         , SBE.sBlockViewers = MapF.fromList blockViewers
+                                         , SBE.sFunctionViewer = MapF.fromList funcViewers
+                                         , SBE.sFunctionSelector = FS.functionSelector (const (return ())) focusedListAttr []
+                                         , SBE.sSymbolicExecutionManager =
+                                           SEM.symbolicExecutionManager (Some symbolicExecutionState)
+                                         }
+          tc0 <- C.newTranslationCache
+          ctxStk <- contextStackFromState ng tc0 ares sesID simState pfcfg
+          let archState = C.ArchState { C.sAnalysisResult = ares
+                                      , C.sContext = ctxStk
+                                      , C.sSymExState = C.singleSessionState symbolicExecutionState
+                                      , C.sIRCache = tc0
+                                      , C.sArchDicts = MapF.fromList dicts
+                                      , C.sUIState = uiState
+                                      }
+          let uiExt = SBC.mkExtension (C.writeChan chan) n0 addrParser "M-x"
+          return C.S { C.sInputFile = Nothing
+                     , C.sLoader = Nothing
+                     , C.sLogStore = mempty
+                     , C.sLogActions = C.LoggingActions { C.sStateLogger = C.logToState chan
+                                                        , C.sFileLogger = Nothing
+                                                        }
+                     , C.sDiagnosticLevel = C.Debug
+                     , C.sEchoArea = C.echoArea 10 (resetEchoArea chan)
+                     , C.sAppState = C.Ready
+                     , C.sNonceGenerator = ng
+                     , C.sKeymap = SBK.defaultKeymap (Just n0)
+                     , C.sArchNonce = n0
+                     , C.sEventChannel = chan
+                     , C.sUIExtension = uiExt
+                     , C.sValueNames = C.initialValueNames (Proxy @arch) simState
+                     , C.sArchState = Just archState
+                     , C.sUIMode = C.SomeUIMode C.SymbolicExecutionManager
+                     }
+        _ -> error "OverrideFrame has no parent frame"
```

Comment:
Is there anything we can share between the override frame and call frame cases? There is quite a lot duplicated here, and it would be unfortunate if it had to stay that way

Reply:
I moved what I could out, but couldn't move these blocks, they need the `arch` and `s` parameters

Reply:
Moved now.